src/runtime/objmodel.cpp

```diff
@@ -212,18 +212,47 @@ extern "C" void my_assert(bool b) {
 }
 
 bool isInstance(Box* obj, BoxedClass* cls) {
-    return isSubclass(obj->cls, cls);
+    BoxedClass* obj_cls = obj->cls;
+#define FAST_CLS_CHECK(_cls, flag)                                                                                     \
+    if (cls == (_cls))                                                                                                 \
+        return PyType_FastSubclass(obj_cls, flag);
+    FAST_CLS_CHECK(BaseException, Py_TPFLAGS_BASE_EXC_SUBCLASS);
+    FAST_CLS_CHECK(type_cls, Py_TPFLAGS_TYPE_SUBCLASS);
+    FAST_CLS_CHECK(int_cls, Py_TPFLAGS_INT_SUBCLASS);
+    FAST_CLS_CHECK(long_cls, Py_TPFLAGS_LONG_SUBCLASS);
+    FAST_CLS_CHECK(str_cls, Py_TPFLAGS_STRING_SUBCLASS);
+    FAST_CLS_CHECK(unicode_cls, Py_TPFLAGS_UNICODE_SUBCLASS);
+    FAST_CLS_CHECK(tuple_cls, Py_TPFLAGS_TUPLE_SUBCLASS);
+    FAST_CLS_CHECK(list_cls, Py_TPFLAGS_LIST_SUBCLASS);
+    FAST_CLS_CHECK(dict_cls, Py_TPFLAGS_DICT_SUBCLASS);
+#undef FAST_CLS_CHECK
+
+    int rtn = PyObject_IsInstance(obj, cls);
+    if (rtn < 0)
+        checkAndThrowCAPIException();
+    return rtn;
 }
 
 extern "C" bool isSubclass(BoxedClass* child, BoxedClass* parent) {
-    // TODO the class is allowed to override this using __subclasscheck__
-    assert(child->tp_mro);
-    assert(child->tp_mro->cls == tuple_cls);
-    for (auto b : static_cast<BoxedTuple*>(child->tp_mro)->elts) {
-        if (b == parent)
-            return true;
-    }
-    return false;
+#define FAST_CLS_CHECK(cls, flag)                                                                                      \
+    if (parent == (cls))                                                                                               \
+        return PyType_FastSubclass(child, flag);
+
+    FAST_CLS_CHECK(BaseException, Py_TPFLAGS_BASE_EXC_SUBCLASS);
+    FAST_CLS_CHECK(type_cls, Py_TPFLAGS_TYPE_SUBCLASS);
+    FAST_CLS_CHECK(int_cls, Py_TPFLAGS_INT_SUBCLASS);
+    FAST_CLS_CHECK(long_cls, Py_TPFLAGS_LONG_SUBCLASS);
+    FAST_CLS_CHECK(str_cls, Py_TPFLAGS_STRING_SUBCLASS);
+    FAST_CLS_CHECK(unicode_cls, Py_TPFLAGS_UNICODE_SUBCLASS);
+    FAST_CLS_CHECK(tuple_cls, Py_TPFLAGS_TUPLE_SUBCLASS);
+    FAST_CLS_CHECK(list_cls, Py_TPFLAGS_LIST_SUBCLASS);
+    FAST_CLS_CHECK(dict_cls, Py_TPFLAGS_DICT_SUBCLASS);
+#undef FAST_CLS_CHECK
+
+    int rtn = PyObject_IsSubclass(child, parent);
```

Comment:
hmm I think the intention of isSubclass is more similar to PyType_IsSubtype (only checks new style classes, doesn't check `__subclasscheck__`) than to PyObject_IsSubclass.